### DISCUSSION TITLE
TS issue after upgrading to v. 7.9 (Type 'ChainablePromiseElement<Promise<Element>>' is not assignable to type 'Promise<Element>'

### DISCUSSION THREAD

**Discussion by bad-white-cat** [webdriverio/webdriverio#7202] [Q&A]
Hello everyone, thank you all for this long-awaited update! 

I tested it a bit and found a TS issue

```
export class Page {
    get container(): Promise<WebdriverIO.Element> {
        return $('div[class^="buttons"]') //warning 1
    }
    get button(): Promise<WebdriverIO.Element> {
        return this.container.$('a.button') // warning 2
    }
}
```
I have the following TS warning here

Warning 1:
```
Type 'ChainablePromiseElement<Promise<Element>>' is not assignable to type 'Promise<Element>'.
  Types of property 'then' are incompatible.
    Type '<TResult1 = Promise<Element>, TResult2 = never>(onfulfilled?: (value: Promise<Element>) => TResult1 | PromiseLike<TResult1>, onrejected?: (reason: any) => TResult2 | PromiseLike<...>) => Promise<...>' is not assignable to type '<TResult1 = Element, TResult2 = never>(onfulfilled?: (value: Element) => TResult1 | PromiseLike<TResult1>, onrejected?: (reason: any) => TResult2 | PromiseLike<...>) => Promise<...>'.
      Types of parameters 'onfulfilled' and 'onfulfilled' are incompatible.
        Types of parameters 'value' and 'value' are incompatible.
          Type 'Promise<Element>' is missing the following properties from type 'Element': uploadToInvisibleField, isMultiremote, elementId, ELEMENT, and 99 more.
```
Warning 2:
```
Property '$' does not exist on type 'Promise<Element>'.
```

And in the test
```
describe('sample spec', () => {
    it('check chaining inside of page object', async () => {
        await browser.url('https://webdriver.io/') 
        const page = new Page()
        await expect(page.button).toHaveText('Get Started')
        await page.button.click() // Property 'click' does not exist on type 'Promise<Element>'.
    });
});
```
Test works okay, though.

Could anyone please advice, how to write typings here? Or maybe there is an issue with them? Has I done anything wrong with update?

**Comment by erwinheitzman** [comment]
You are trying to assign your own types which is causing issues. Don't assign types manually and let TypeScript handle that instead.

**Reply by NavyMint** [reply]
so we should do this way? Sorry for disturbing you in all threads :) We have also same problem in page objects and I decided to try 
![image](https://user-images.githubusercontent.com/26191692/127508225-c168fe8a-964a-449b-9e99-185f3c2a9f20.png)

![image](https://user-images.githubusercontent.com/26191692/127508282-056cedd5-e727-4d82-8682-8336aa10ddd4.png)

**Reply by bad-white-cat** [reply]
> You are trying to assign your own types which is causing issues. Don't assign types manually and let TypeScript handle that instead.

Hello Erwin, thank you for your answer.

**Reply by NavyMint** [reply]
@bad-white-cat sorry, is it working for you now?

**Reply by bad-white-cat** [reply]
@NavyMint I think like that 

```
export class Page {
    get container() {
        return $('div[class^="buttons"]')
    }
    get button() {
        return this.container.$('a.button') 
    }
}
```
It makes sense as these types are all inferrable 😶

P.S. Tests were not broken, in my case it was just TS problem

**Reply by NavyMint** [reply]
@bad-white-cat got it, thanks :)
in my case its also just a TS problem, but unfortunately we have a es-lint rule to deprecate missed returning types and its a bad idea to suppress it everywhere or delete rule at all, so we need to use something 
@erwinheitzman can you please suggest can we use any workaround in this case?

**Reply by erwinheitzman** [reply]
You will have to replace all the return types with the correct (new) type but you will run into issues if the type would ever change again in the future

**Reply by NavyMint** [reply]
@erwinheitzman  okey, then we have again a probmlem with custom commands :(

if we will rework classes to take as parameters element with a new Chainable type, it will throw an error when Im trying to pass element searched by custom command, coz it returns Promise\<Element\> and vise versa

and we need it to work in both cases

![image](https://user-images.githubusercontent.com/26191692/127519772-ae8420df-26be-426d-ad47-ce3d7bee4532.png)
![image](https://user-images.githubusercontent.com/26191692/127520192-a5730e91-eaaa-4edd-91f5-af6d3bf53d11.png)

![image](https://user-images.githubusercontent.com/26191692/127519932-b9a4a6f8-8f45-4387-9e1b-5bb5f48cd831.png)

**Reply by erwinheitzman** [reply]
@NavyMint I am sorry but as I said, the problem lies in assigning the wrong type (`Promise<Element>` versus the `ChainablePromiseElement`).

**Reply by NavyMint** [reply]
@erwinheitzman yeah, sorry, I got, but it looks like a bug if we can't use simultaneously custom commands and general elements with their original type and not to break typings in ts project, this way its also not ok 

![image](https://user-images.githubusercontent.com/26191692/127525897-56c2d833-d967-40d1-9086-597fc57fc0af.png)
![image](https://user-images.githubusercontent.com/26191692/127525998-0e2d8d9b-9d26-4dee-adba-b7c085387c96.png)
![image](https://user-images.githubusercontent.com/26191692/127526038-e88e4005-12e1-4aab-bcc6-db4565b39c57.png)

**Reply by christian-bromann** [reply]
@NavyMint one tip that makes it easier for others: paste code like this:
```ts
type foo: number = 42
```
rather than code pictures as we can copy/paste this into our IDEs.

Then regarding your problem:
I suggest to do it this way:

```ts
    class PageObject {
        constructor (public element: WebdriverIO.Element) {}
    
        get subElem (): ChainablePromiseElement<Promise<WebdriverIO.Element>> {
            return this.element.$('subElem')
        }
    }
    
    const elem = await $('foo')
    const page = new PageObject(elem)
    const text = await page.subElem.getText()
```

**Reply by christian-bromann** [reply]
I will update the docs to reflect this

**Reply by NavyMint** [reply]
@christian-bromann thank you for response and example, we are really appriciate all your support guys.
I've tried your suggestion and in this way it works fine, but in your example we need to pass awaited elements, and its not always convenient. And when I try to do same but with promise in constructor -> its not working simultaneously for 2 cases (custom command and simple element searched by $)

My example Im strugglin' with 

**class**

```
class ButtonWebElement {
    constructor(public element: Promise<WebdriverIO.Element>) {}
...
}
```

**test case:**

```
const el = new ButtonWebElement((await $('el')).dataId('el'));
const el2 = new ButtonWebElement($('el'));  **<= this one is not working in such approach**
```
`error for el2 ->> Argument of type 'ChainablePromiseElement<Promise<Element>>' is not assignable to parameter of type 'Promise<Element>'.`

**custom command type** 

```
declare namespace WebdriverIO {
    interface Element {
        dataId: (selector: string) => Promise<Element>;
    }
}
```

in this case we can pass to class element searched by custom command and can't pass elements searched by $ . Its not too convenient to work with awaited elements, coz in this case we need to rework getters to async functions in page objects, due to await usage

**Reply by christian-bromann** [reply]
You can do the same without awaiting the promise:

```ts
    class PageObject {
        constructor (public element: ChainablePromiseElement<Promise<WebdriverIO.Element>>) {}
    
        get subElem (): ChainablePromiseElement<Promise<WebdriverIO.Element>> {
            return this.element.$('subElem')
        }
    }
    
    const page = new PageObject($('foo'))
    const text = await page.subElem.getText()
```

as @erwinheitzman mentioned, it is just a matter how you define your types.

**Reply by NavyMint** [reply]
okey, If I will do this way then simple searched element will work, but element found by custom command and passed to class - no, 
then we need to update custom command return type also to ChainablePromiseElement

**from**
```
declare namespace WebdriverIO {
    interface Element {
        dataId: (selector: string) => Promise<Element>;
    }
}
```

**to** 

```
interface Element {
        dataId: (selector: string) => ChainablePromiseElement<Promise<Element>>;
    }
```

and after changing return type of custom command to ChainablePromiseElement Im getting an error that my custom command doesn't exists on type Element

error 
`Property 'dataId' does not exist on type 'Element'`

**Reply by NavyMint** [reply]
as I said I need to find a ballance to reach 3 goals at the same time:
1) Class should have as parameter Promise element
2) We should be able to pass to class element searched by $()
3) We should be able to pass to class element searched by custom command

and from suggestions above I can or await all elements and pass both types to class, or to have promise in class and pass only elemenets searched by $ and not by custom command, or by custom command and not by $()

**Reply by erwinheitzman** [reply]
It works just fine for me

```js
// test.spec.ts
import { ChainablePromiseElement } from 'webdriverio';
import loginPage from '../pageobjects/login.page';
import securePage from '../pageobjects/secure.page';

class ButtonWebElement {
    constructor(public element: ChainablePromiseElement<Promise<WebdriverIO.Element>>) {}

    async getPassword() {
        return (await this.element).$('#password');
    }
}

describe('My Login application', () => {
    it('should login with valid credentials', async () => {
        loginPage.open();

        browser.addCommand('example', (selector: string) => {
            return $(selector);
        }, true);
        
        const body = await $('body');
        await (await body.example('#username')).setValue('invalid');

        const el1 = new ButtonWebElement($('body'));
        await (await el1.getPassword()).setValue('some value');
        
        const el2 = new ButtonWebElement(body.example('#content'));
        await (await el2.getPassword()).setValue('some other value');
        
        await loginPage.btnSubmit.click();
        await securePage.flashAlert.waitForDisplayed();
    });
});
```

```js
// wdio.d.ts
declare namespace WebdriverIO {
    interface Element {
        example: (selector: string) => ChainablePromiseElement<Promise<WebdriverIO.Element>>
    }
}
```

**Reply by NavyMint** [reply]
Very strange thing, same code not working for me, copy-pasted, but got same error as I have in my case
`Property 'example' does not exist on type 'Element'`

**Reply by erwinheitzman** [reply]
@NavyMint did you use the latest version?

**Reply by NavyMint** [reply]
yep, overall thats all starts to happening after update to 7.9

**Reply by erwinheitzman** [reply]
@NavyMint I can't do more then acually share the setup: https://github.com/erwinheitzman/wdio-typings

**Reply by NavyMint** [reply]
@erwinheitzman thanks a lot for all your help and time, I will take a look, for now we decided to go with 'any' in wdio.d.ts, its also have some pitfalls but anyway, we will try :))

**Reply by NavyMint** [reply]
@erwinheitzman I found our problem, with this property "skipLibCheck": true in tsconfig everything works as a charm )))yay, thanks, your repo really helped a lot

**Reply by versedi** [reply]
@NavyMint I wouldn't call setting skipLibCheck to true a solution. 

According to the docs:
 



Skip Lib Check - skipLibCheck

> Skip type checking of declaration files.

> This can save time during compilation at the expense of type-system accuracy. For example, two libraries could define two copies of the same type in an inconsistent way. Rather than doing a full check of all d.ts files, TypeScript will type check the code you specifically refer to in your app’s source code.

> A common case where you might think to use skipLibCheck is when there are two copies of a library’s types in your node_modules. In these cases, you should consider using a feature like yarn’s resolutions to ensure there is only one copy of that dependency in your tree or investigate how to ensure there is only one copy by understanding the dependency resolution to fix the issue without additional tooling.


You might run into troubles later on in the future by changing this option to true. 

@erwinheitzman any particular reason to have `skiLibCheck` flag enabled? 

Came here from Google result, I'm having similar issue as OP and `skipLibCheck` can't be changed in my case.

**Reply by NavyMint** [reply]
@versedi we've already got that itss not a best approach and decisided to stick to solution when we just specifying type any in wdio.d.ts
